### PR TITLE
Introduce NamedArguments rule

### PR DESCRIPTION
**Pull request by sowmyav24**
Fixes [#1007](https://github.com/detekt/detekt/issues/1007)

Changes done:

- Introduce `NamedArguments` rule which enforces function invocation to have named parameters when they cross the threshold of number of non-named parameters.
- Currently, the default is set to `3`.

### CODE REVIEW

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Severity
+import io.gitlab.arturbosch.detekt.api.ThresholdRule
+import org.jetbrains.kotlin.psi.KtCallExpression
+
+/**
+ * Reports functions invocations which have more parameters than a certain threshold and are not named.
+ *
+ * @configuration threshold - number of non-named parameters allowed in function invocation (default: `3`)
```
Comment (cortinico) [inline detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt:15]:
```suggestion
 * Reports functions invocations which have more parameters than a certain threshold and are all not named.
 *
 * @configuration threshold - number of parameters that triggers this inspection (default: `3`)
```

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ */
+class NamedArguments(
+    config: Config = Config.empty,
+    threshold: Int = DEFAULT_FUNCTION_THRESHOLD
+) : ThresholdRule(config, threshold) {
+
+    override val issue = Issue(
+        "NamedArguments", Severity.Maintainability,
+        "Function invocation with more number of parameters must be named.",
```
Comment (cortinico) [inline detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt:24]:
```suggestion
        "Function invocation with more than $threshold parameters must all be named.",
```

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        val valueArguments = expression.valueArguments
+        if (valueArguments.size > functionInvocationThreshold && valueArguments.any { !it.isNamed() }) {
+            report(CodeSmell(issue, Entity.from(expression), issue.description))
+        } else {
+            super.visitCallExpression(expression)
+        }
+    }
+
+    companion object {
+        const val THRESHOLD = "threshold"
```
Comment (cortinico) [inline detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt:43]: I don't think you need this. You should be able to use `.threshold` from the upper class.

**Review by cortinico — COMMENTED** [review]
Thanks for your contrib @sowmyav24 

Please clarify the documentation as the semantic is a bit unclear right now. The threshold applies to the whole number of parameters, and not just to the named parameters.

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt (hunk 3 above)
Comment (sowmyav24) [reply detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt]: Removed it.

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt (hunk 2 above)
Comment (sowmyav24) [reply detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt]: Made the changes.

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt (hunk 1 above)
Comment (sowmyav24) [reply detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt]: Sure, I have incorporated the above.

**Review by BraisGabin — APPROVED** [review]

**Review by cortinico — APPROVED** [review]
LGTM 👍

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import io.gitlab.arturbosch.detekt.api.Debt
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Severity
+import io.gitlab.arturbosch.detekt.api.ThresholdRule
+import org.jetbrains.kotlin.psi.KtCallExpression
+
+/**
+ * Reports function invocations which have more parameters than a certain threshold and are all not named.
+ *
```
Comment (schalkms) [inline detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt:14]: A compliant and non-compliant code sample helps the user to understand the purpose of the rule. I think it's also helpful here.

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+) : ThresholdRule(config, threshold) {
+
+    override val issue = Issue(
+        "NamedArguments", Severity.Maintainability,
+        "Function invocation with more than $threshold parameters must all be named",
+        Debt.FIVE_MINS
+    )
+
+    override fun visitCallExpression(expression: KtCallExpression) {
+        val valueArguments = expression.valueArguments
```
Comment (schalkms) [inline detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt:29]: What happens when calling constructors? Shall a test be added?

**Review by schalkms — COMMENTED** [review]
Thanks for the contribution! Please see my attached comments. I'm not sure whether there should be a bigger test coverage.

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt (hunk 4 above)
Comment (sowmyav24) [reply detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt:14]: @schalkms Added the code snippets here 👍

File: detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt (hunk 5 above)
Comment (sowmyav24) [reply detekt-rules-complexity/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NamedArguments.kt:41]: The rule gets triggered. Have added tests for them

**Comment by schalkms** [comment]
@cortinico & @BraisGabin should this rule also report constructors or just functions? What's your take on this?

**Comment by BraisGabin** [comment]
I think that it should report constructors. But maybe we could have two different threshold.

**Comment by cortinico** [comment]
> @cortinico & @BraisGabin should this rule also report constructors or just functions? What's your take on this?

Seems like it works also for ctors.

> I think that it should report constructors. But maybe we could have two different threshold.

Don't have a strong opinion on this actually. Considered that the rule is not enabled by default, I'd be fine having a single threshold for now and adding another one in the future if needed.

**Pull request merged**